### DISCUSSION TITLE
How do you deploy Livewire?

### DISCUSSION THREAD

**Discussion by robsontenorio** [livewire/livewire#7131] [1. Help] [answered]
Hi, guys!

How looks like your deployment script?

Currently I am doing this . Is there anything to improve here?

```
php artisan config:cache
php artisan view:cache
php artisan route:cache
```

If deploy happens when you are logged in, some components glitches. Like `wire:loading`  triggered improperly.  I don't know exactly how to reproduce this. Am I missing something ?

```sh
#!/usr/bin/zsh

echo '------ Starting deploy tasks  ------'

cp .env.example .env
composer install --prefer-dist --no-interaction --no-progress --ansi

yarn install
yarn build

php artisan migrate --seed --force

php artisan config:cache
php artisan view:cache
php artisan route:cache
php artisan icons:cache

echo '------ Deploy completed ------'
```

```blade
<!DOCTYPE html>
<html lang="{{ str_replace('_', '-', app()->getLocale()) }}">

<head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, viewport-fit=cover">
    <meta name="csrf-token" content="{{ csrf_token() }}">
    <title>{{ $title ?? config('app.name') }}</title>
    <link rel="shortcut icon" type="image/x-icon" href="{{ asset('/favicon.ico') }}">

     @vite(['resources/css/app.css'])

</head>
```

```js
import { defineConfig } from 'vite';
import laravel from 'laravel-vite-plugin';

export default defineConfig({
    plugins: [
        laravel({
            input: ['resources/css/app.css', 'resources/js/app.js'],
            refresh: true,
        }),
    ],
});
```

**Comment by utqinadhif** [comment]
try to add x-cloak on HTML and add x-cloak { display: none } to style.css

**Reply by robsontenorio** [reply]
@utqinadhif how looks like your deploy script ?

**Reply by lrljoe** [reply]
@robsontenorio <- how about you share your deployment script, and people can give you tips & hints?  I can guarantee that 90%+ of my deployment scripts will be completely irrelevant to you.

Asking people to randomly share scripts, with zero context, is **not** going to help you in the slightest.

Also share your frontend approach, are you using Tailwind, Bootstrap, or something else entirely?

You say that you're seeing glitching when deploying, I'm assuming you're doing a proper build, and not using the Vite development server?

**Reply by robsontenorio** [reply]
@lrljoe it is tailwind and vite. Just updated my original comment with more context.

I would say your deploy script would be relevant for me , once I can spot something important I am missing.

**Reply by lrljoe** [reply]
My deployment pipeline makes various calls out to spin up Azure/AWS resources as required, so would probably add to the confusion!  Asking everyone for their deployment scripts, then trying to pluck lines out, will end in disaster, and numerous more queries.

Looking at your deployment process, I think you're 100% missing a trick (or 3!).

Keep in mind the difference between:
1) Updating a site from a repo
2) Updating dependencies
3) Building a site from scratch.

Your build approach is a No.3 approach.  You're creating a completely new instance of your site.  So, yeah, you're going to see something glitching when you completely wipe out a site and replace it with a new instance, not sure what you're expecting there?

Key take-aways:
Your "deployed" site shouldn't ever need any yarn/npm build processes etc, that's what your build pipeline is supposed to be doing! 
So your "Deploy" branch (or whatever you like to call it) should have a trusted, compiled set of your compiled/built JS scripts, so that you're not having to add all of the "junk" (for want of a better word) development requirements into your production environment.

You're copying .env.example to .env <- I'm assuming that you're then injecting the proper environmental variables in somewhere else along your build process?

You're forcing a "fresh" deployment to your database?  Is this really what you want for your deployment?

What're you trying to achieve with your deployment script, are you trying to build new instances of your site, trying to update it, or what?

**Reply by robsontenorio** [reply]
@lrljoe  Sorry for the typo it is `php artisan migrate --seed --force`

It is a docker deployment .

So, I am running those `composer` and `yarn` because all assets will be present on final docker image to be deployed . So when build is done the image is replaced on the fly at cloud provider.

**Reply by lrljoe** [reply]
Bit of an odd approach to deploying to the cloud to be honest!

You say you're preparing an image for the cloud.  Using docker in production at scale, is really almost unheard of, and even Docker themselves don't really recommend that as an approach...  
Even then, why would you be re-seeding a database that is in production?

At the moment, your deployment process appears to be fundamentally flawed (to me at least), and I can't really advise further without really understanding what your use case is.

Your original issue is that things "glitch" when you deploy.  Now, you're deploying a fresh container, to which I can only say - things will glitch.  If you're swapping out entire VMs (by re-deploying a Docker image - you're doing just that), then things will always glitch, unless you switch out to an enterprise approach, with shared cache disks etc, which goes **way** beyond what you can expect in a reply to a discussion on here.

**Comment by dakira** [comment, marked as answer]
My deployment looks like this:

- I've got my app in a Docker container (based on the serversideup/docker-php images).
- the image build process adds composer dependencies, optimizes autoload and builds assets
- it also does all caching that doesn't depend on runtime
- The container has a laravel-automations startup-script that handles the rest (migrations, caching, storage linking...).

The images are built by CI and certain tags cause them to be deployed to certain environments automatically.
